File: act_reto.py
import control as co
import matplotlib.pyplot as plt
import numpy as np

# ------------------------------------
# PLANTA (2º orden, estable)
# Polos ≈ -5000 ± j3000
# ------------------------------------
g1 = co.tf([1e7], [1, 10000, 3.4e7])

# ------------------------------------
# Proporcional
# ------------------------------------
p = co.tf([3], [1])

# ------------------------------------
# Integral (misma forma que MATLAB)
# ------------------------------------
i  = co.tf([1], [10])
gi = 1000
ig = co.series(gi, i)

# ------------------------------------
# Derivativo (misma forma que MATLAB)
# ------------------------------------
d  = co.tf([10], [1])
gd = 0.001
dg = co.series(gd, d)

# ------------------------------------
# PID en paralelo
# ------------------------------------
pi  = co.parallel(p, ig)
pid = co.parallel(pi, dg)

# ------------------------------------
# Serie con la planta
# ------------------------------------
g = co.series(pid, g1)

# ------------------------------------
# Realimentación unitaria
# ------------------------------------
Gs = co.feedback(g, 1)

# ------------------------------------
# Respuesta al escalón
# ------------------------------------
t, y = co.step_response(Gs)

plt.figure(figsize=(7,6))

plt.subplot(2,1,1)
plt.plot(t, y)
plt.grid(True)
plt.title("Step Response")

# ------------------------------------
# Mapa polos-ceros
# ------------------------------------
plt.subplot(2,1,2)
co.pzmap(Gs, plot=True, grid=True)
plt.title("Pole-Zero Map")

plt.tight_layout()
plt.show()

File: ACT_RETO.m
%% SISTEMA QUE DA LA GRAFICA QUE MOSTRASTE

% Planta de 2º orden con polos complejos ESTABLES
% Polos ≈ -5000 ± j3000
g1 = tf([1e7], [1 10000 3.4e7]);

%% Controlador (ganancia equivalente, estable)
p  = 3;

i  = tf([1],[10]);
gi = 1000;

d  = tf([10],[1]);
gd = 0.001;

%% Series
ig = series(gi,i);
dg = series(gd,d);

%% Paralelo
pi  = parallel(p,ig);
pid = parallel(pi,dg);

%% Serie total
g = series(pid,g1);

%% Realimentacion unitaria
Gs = feedback(g,1);

%% Graficas
figure

subplot(2,1,1)
step(Gs)
grid on
title('Step Response')

subplot(2,1,2)
pzmap(Gs)
grid on
title('Pole-Zero Map')
